Fix part of #1366 : create continuous job for creator dashboard statistics

526avijitgupta opened the pull request:
Tasks to be completed as a part of this PR:

- [x] Model Changes
- [x] Modify existing job for aggregation of dashboard related stats
- [x] Modify and add tests for the job

seanlip commented on core/storage/user/gae_models.py:
```diff
@@ -148,6 +148,20 @@ class UserStatsModel(base_models.BaseMapReduceBatchResultsModel):
     # The impact score.
     impact_score = ndb.FloatProperty(indexed=True)
 
+class UserDashboardStatsModel(base_models.BaseMapReduceBatchResultsModel):
```
Any reason not to reuse UserStatsModel above?

526avijitgupta replied:
I was a bit uncertain about this, but then chose to create a new model based on the fact that these stats are shown in a user's dashboard (in contrast to profile page).

seanlip replied:
I lean towards reusing, actually. Everything in both models is updated using batch jobs, so it seems reasonable to bring them all together. The impact score isn't large so I think there's no need to worry about retrieving too much unneeded info, and the UIs of the dashboard and profile page may well change in the future to highlight more of these stats.

526avijitgupta replied:
@seanlip I did follow your suggestion and reused the existing model instead of creating a new one. But the [instrcutions here](https://github.com/oppia/oppia/wiki/Calculating-statistics#finding-the-right-job) clearly state that no 2 jobs should be updating the same model.

Thus I'm forced to think that we might have to:
1. either revert back to using different models,
2. or re-write the existing impact aggregator job to be able to make the necessary aggregations for average ratings and number of plays as well.
I'm not sure if  (2) is even feasible -- your take on this ?

seanlip replied:
Actually, the existing impact aggregator job makes use of ratings and plays in its calculation. It seems to me like you should be able to reuse it, and get it to calculate more stuff?

526avijitgupta commented:
@seanlip I was a bit uncertain about reusing the existing (UserImpact) job to calculate more stats since the user impact is measured for all _contributors_ of an exploration. On the other hand, 'average ratings' and 'total plays' have to be calculated only for the _owners_ of an exploration. 
I have tried to accommodated the required functionality into a single job, PTAL. The tests for this are yet to be updated. (Though I've manually tested this job by creating for a single exploration with non-zero values for both average ratings and total plays).

seanlip commented:
Someone needs to review this in more detail ( @AllanYangZhou ?) but the general approach looks good to me!

seanlip commented on core/domain/user_jobs_continuous.py:
```diff
@@ -346,19 +349,50 @@ def map(item):
+                # Get num starts (total plays) for the exploration
+                num_starts = statistics['start_exploration_count']
+                exploration_data.update({
+                    'exploration_total_plays': num_starts,
```
maybe call this total_plays_for_owned_exp, average_rating_for_owned_exp (just to be clear)

526avijitgupta replied:
Resolved in https://github.com/oppia/oppia/pull/1891/commits/38191eb16eae1190b586fda3666e99815e633c79

seanlip commented on core/domain/user_jobs_continuous.py:
```diff
@@ -317,12 +322,10 @@ def map(item):
-            return
-        average_rating = total_rating / sum(item.ratings.itervalues())
+        average_rating = (
+            total_rating / sum(item.ratings.itervalues())
```
Not sure it's needed to calculate the sum twice. Maybe calculate it just once, store it in sum_of_ratings, and refer to it in this line and the next?

526avijitgupta replied:
Done

seanlip commented on core/domain/user_jobs_continuous.py:
```diff
@@ -346,19 +349,50 @@ def map(item):
+                num_starts = statistics['start_exploration_count']
+                exploration_data.update({
+                    'exploration_total_plays': num_starts,
+                    'exploration_average_rating': average_rating
```
Actually I think this should be None if there are no ratings, and it should just not be counted in the calculation. What do you think?

"No ratings" is different from "rating is 0".

526avijitgupta replied:
Done

seanlip commented on core/domain/user_jobs_continuous.py:
```diff
@@ -346,19 +349,50 @@ def map(item):
+        # Find the average of all average ratings
+        ratings = [value['exploration_average_rating'] for value in values
+                   if value.get('exploration_average_rating')]
+        average_ratings = (sum(ratings) / float(len(ratings))
```
Again, careful about the None case.

526avijitgupta replied:
Done

seanlip commented:
@526avijitgupta -- I did a fuller review pass, since @AllanYangZhou is a bit busy with the RTE. Overall it looks good! Can you please get the tests in?

Thanks!

526avijitgupta commented:
@seanlip . Thanks for the review. I'll address the reviewer comments soon and start writing the tests from tomorrow.

526avijitgupta commented on core/storage/user/gae_models.py:
```diff
@@ -152,10 +152,10 @@ class UserStatsModel(base_models.BaseMapReduceBatchResultsModel):
     # The average of average ratings of all explorations.
-    average_ratings = ndb.FloatProperty(indexed=True)
+    average_ratings = ndb.FloatProperty(default=None, indexed=True)
 
```
@seanlip I have added default values to the model itself to make things clear.

seanlip replied:
Hm, actually, I think you don't need them. None is automatically the default value if no default arg is specified -- and, for total_plays, None tells us it's never been calculated, which is a different meaning from 0. So I would recommend just not having the default arg for either.

Thanks for flagging this!

526avijitgupta replied:
resolved by https://github.com/oppia/oppia/pull/1891/commits/1ff6be6f85269c0b6f8f76ff16ee576420b39d5e

526avijitgupta commented on core/domain/user_jobs_continuous.py:
```diff
@@ -317,12 +322,10 @@ def map(item):
-        if not sum(item.ratings.itervalues()):
-            return
-        average_rating = total_rating / sum(item.ratings.itervalues())
+        average_rating = (
```
Resolved

526avijitgupta commented:
@seanlip Made changes to map and fixed failing tests of impact score calculation. As per our earlier discussion, testing for average_ratings and num_plays will be done end-to-end (inside DashboardHandlerTest() in dashboard_test.py) which will require services/controllers to be written first, I guess ?
If so, I'll remove the testing part from the "tasks to be done as a part of PR" and instead open a new issue for the same. That will also mean that the tasks for this PR have been completed. It needs to be reviewed.

@wxyxinyu - I've assigned you to this PR since I would prefer that you review it!

seanlip commented:
Hi @526avijitgupta, thanks for the update! To clarify, when I said end-to-end, I just meant: try to do it without mocking stuff. You shouldn't need to wait for additional services/controllers to be written; it should be possible to test the surface of what you've already written. Tests should be written before merge into develop.

526avijitgupta commented:
@seanlip If I understand correctly then, events will be triggered (playing an exploration, rating an exploration) to write to data model, but the model's value will be read directly (only read, no write) to check for correct values ?
Also, are there any similar examples to this approach in the current codebase (Model is written through events, and read directly to verify for correct values) ? (Did a quick search - couldn't find any).

526avijitgupta commented:
Yes it sounds good. thanks!

wxyxinyu commented on core/domain/user_jobs_continuous.py:
```diff
@@ -258,18 +258,18 @@ def reduce(key, stringified_values):
 
-class UserImpactAggregator(jobs.BaseContinuousComputationManager):
+class UserStatsAggregator(jobs.BaseContinuousComputationManager):
     """A continuous-computation job that computes the impact score
```
Update the docstring: "impact score" -> "user stats for creator dashboard"?

526avijitgupta replied:
Done

wxyxinyu commented on core/domain/user_jobs_continuous_test.py:
```diff
@@ -705,6 +713,7 @@ def test_standard_user_impact_calculation_one_exploration(self):
 
         # Verify that the impact score matches the expected.
         self._run_computation()
+        # self.process_and_flush_pending_tasks()
```
remove commented out code

526avijitgupta replied:
Done

wxyxinyu commented on core/storage/user/gae_models.py:
```diff
@@ -136,7 +136,13 @@ class UserRecentChangesBatchModel(base_models.BaseMapReduceBatchResultsModel):
+    Impact scores are calculated over explorations for which a user
+    is listed as a contributor
+
+    The impact score for a particular user, where impact is defined as:
```
suggest "The impact score for a particular user is defined as:"

526avijitgupta replied:
Done

wxyxinyu commented:
Code looks good generally, I have a couple of small comments.

Regarding the tests, the purpose of unit tests here is to make sure that the job is running correctly, i.e. the process from reading from the models to writing the models is right. So to achieve this, you should simulate the events by making the appropriate backend calls (e.g. creating users and explorations, and simulating ratings and finishing explorations). It does not need to involve the frontend controller code at all. Does that make sense?

526avijitgupta commented:
@seanlip @wxyxinyu I've written an integration test which runs the ModifiedUserStatsAggregator to compute and assert the values of UserStatsModel. I was thinking that it would be pretty helpful if you could take a quick review of this basic implementation, and see whether the approach is correct.

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +71,66 @@ def test_logged_in_notifications_dashboard(self):
 
 
+class DashboardStatisticsTest(test_utils.GenericTestBase):
+    # TODO: Replace the model reads in each of the tests inside this class by
```
TODO(526avijitgupta):

526avijitgupta replied:
Done

seanlip commented:
Yep, looks reasonable to me!

526avijitgupta commented:
Thanks @seanlip for the quick review !

526avijitgupta commented:
@wxyxinyu @seanlip @AllanYangZhou Please review.

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
 
 
+class DashboardStatisticsTest(test_utils.GenericTestBase):
+    # The model reads in each of these tests inside this class will not be
```
I don't think you need this comment. It's relevant "in the moment", but probably won't make a lot of sense to a reader after several months, since there isn't a general presumption that model reads need to be replaced.

526avijitgupta replied:
Removed

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+    EXP_ID_2 = 'exp_id_2'
+    EXP_TITLE_2 = 'Exploration title 2'
+
+    USER_SESSION = 'session1'
```
SESSION_ID or USER_SESSION_ID

526avijitgupta replied:
Used latter

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+                rating = 1
+        self.process_and_flush_pending_tasks()
+
+    def _start_computation_job(self):
```
_run_computation_job()?

seanlip replied:
Actually: _run_user_stats_aggregator_job()? More specific.

526avijitgupta replied:
Done

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+            0)
+        self.process_and_flush_pending_tasks()
+
+    def _start_job_to_compute_num_plays(self):
```
start -> run

seanlip replied:
_run_stats_aggregator_job()

526avijitgupta replied:
Done

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+            0)
+        self.process_and_flush_pending_tasks()
+
+    def _get_default_value_of_impact_score(self):
```
Can't this and the next one just be class-level constants?

526avijitgupta replied:
Replaced by class level constants.

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+            rating_services.assign_rating_to_exploration(
+                user_id, exp_id, rating)
+            rating = rating + 1
+            if rating > 5:
```
This wrapping seems a little odd, especially in a utility method which is supposed to be straightforward. I think it would be much clearer to just pass in a list of the ratings you want to add -- e.g. _rate_exploration(exp_id, [4]) to give a single rating of 4.

526avijitgupta replied:
Done

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+
+        self._start_computation_job()
+        user_model = user_models.UserStatsModel.get(
+            self.owner_id_1, strict=False)
```
why strict=False?

526avijitgupta replied:
Oops, copy-paste error. Fixed now.

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+        self.assertEqual(len(response['explorations_list']), 1)
+
+        exp_id = self.EXP_ID_1
+        rating_start = 4
```
Here and elsewhere: you don't need this to be a separate variable, especially since you don't use it anywhere else. Just inline it with the method call.

526avijitgupta replied:
Done

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+        self._rate_exploration(exp_id, 3, rating_start)
+
+        self._start_computation_job()
+        user_model = user_models.UserStatsModel.get(
```
Should you be checking for user 2 as well?

526avijitgupta replied:
I don't think that checking for user 2 will be required here, since user 2 is nowhere involved in this particular test. What do you think ?

seanlip replied:
Oh sorry, I put this in the wrong place. I think I meant for it to go in test_stats_for_single_exploration_with_multiple_owners().

526avijitgupta replied:
Already fixed. See latest commit

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -66,6 +75,349 @@ def test_logged_in_notifications_dashboard(self):
+        self._rate_exploration(exp_id_2, 1, rating_start_2)
+
+        self._start_computation_job()
+        user_model = user_models.UserStatsModel.get(
```
Ditto -- no check for the other user?

526avijitgupta replied:
Done

seanlip commented on core/domain/user_jobs_continuous_test.py:
```diff
@@ -595,14 +596,17 @@ class UserImpactAggregatorTest(test_utils.GenericTestBase):
             self.EXP_ID_1: {
                 'complete_exploration_count': (
                     self.num_completions[self.EXP_ID_1]),
+                'start_exploration_count': (
```
Why do you need to add these?

526avijitgupta replied:
This is accessed inside user_jobs_continuous: https://github.com/526avijitgupta/oppia/blob/ad7d7fe78cb809a9c5afa135fa1d786b50e3a1d8/core/domain/user_jobs_continuous.py#L388

So, on removing this it throws me keyError because then the mocked data does not contain `start_exploration_count` key!

seanlip replied:
Ah, got it. OK then:)

seanlip commented on core/domain/user_jobs_continuous_test.py:
```diff
@@ -595,14 +596,17 @@ class UserImpactAggregatorTest(test_utils.GenericTestBase):
-        super(UserImpactAggregatorTest, self).setUp()
+        super(UserStatsAggregatorTest, self).setUp()
         self.num_completions = defaultdict(int)
+        self.num_starts = defaultdict(int)
```
Why is this needed?

526avijitgupta replied:
(See comment above)

526avijitgupta commented:
@seanlip Addressed all review comments.
Also, I realized that map function did not consider the case when a creator of an exploration adds you as an owner (but you have not contributed to the exploration at all) - since previously the owner_ids were being checked out of contributor_ids. Fixed this too.

seanlip commented:
Thanks! Responded to your questions.

seanlip commented on core/controllers/dashboard_test.py:
```diff
@@ -16,13 +16,22 @@
+from core.domain import stats_jobs_continuous_test
 from core.domain import user_jobs_continuous
+from core.domain import user_jobs_continuous_test
 from core.tests import test_utils
```
nit: alphabetize -- put this after core.platform

526avijitgupta replied:
Ok, will fix. 

Just did a quick google search and found that the latest version of pylint now detects that the imports are in alphabetic ordering ->  [SO Answer](http://stackoverflow.com/a/35213338/4135178)

Should we consider updating our pylint to the latest version ?

526avijitgupta replied:
Okay research a bit more and file an issue for this.

526avijitgupta replied:
Fixed the ordering.

seanlip commented:
LGTM! Saw one last trivial thing, but that's it.

526avijitgupta commented:
@seanlip Addressed final review comment. PTAL.
About updating pylint to handle import-ordering well - We already seem to be using an updated version of pylint. Also, the [discussion here](https://github.com/PyCQA/pylint/issues/692#issuecomment-163178642) implies that the import-ordering check fails in some cases. Thus, not investigating further into this -- leaving for now.

seanlip commented:
OK, sgtm.

Also, LGTM! Thanks @526avijitgupta! I've delayed changing the label because I think @wxyxinyu is also currently taking a look at it.

526avijitgupta commented:
okay. I'll continue my work on writing services for this by the time. Thanks for letting me know!

wxyxinyu commented:
lgtm too. Feel free to merge.

526avijitgupta commented:
Thanks @wxyxinyu ! Merging now.
